File: Python3/프로그래머스/4/42897. 도둑질/도둑질.py
def solution(money):
    dp1 = [ 0 for _ in range(len(money))] #첫 집을 턴 경우
    dp1[0] = money[0]
    dp1[1] = money[0]
    
    dp2 = [0 for _ in range(len(money))] #첫 집을 털지 않은 경우
    dp2[1] = money[1]
    
    for i in range(2, len(money)-1):
        dp1[i] = max(dp1[i-2] + money[i], dp1[i-1])
    
    for i in range(2, len(money)):
        dp2[i] = max(dp2[i-2]+money[i], dp2[i-1])
        
    answer = max(dp1[len(money)-2], dp2[len(money)-1])
    
    return answer
        
        
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
#     answer = 0
#     dp1 = [0 for _ in range(len(money))]

#     dp1[0] = money[0]
#     dp1[1] = money[0]
#     for i in range(2, len(money)-1):
#         dp1[i] = max(dp1[i-2] + money[i], dp1[i-1])
    
#     dp2 = [0 for _ in range(len(money))]
#     dp2[0] = 0 #첫집을 털지 않은 경우
#     dp2[1] = money[1]
    
#     for i in range(2,len(money)):
#         dp2[i] = max(dp2[i-2] + money[i], dp2[i-1])
    
#     answer = max(dp1[len(money)-2],dp2[len(money)-1])
#     return answer

File: C++/프로그래머스/4/42897. 도둑질/도둑질.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> money) {
    int money_length = money.size();
    vector<int> dp1(money_length, 0); //첫 집을 턴 경우
    vector<int> dp2(money_length, 0); //첫 집을 털지 않은 경우
    
    dp1[0] = money[0];
    dp1[1] = money[0];
    
    dp2[1] = money[1];
    
    for (int i = 2; i < money_length - 1; i++) {
        dp1[i] = max(dp1[i-2] + money[i], dp1[i-1]);
    }
    
    for (int i = 2; i < money_length; i++) {
        dp2[i] = max(dp2[i-2] + money[i], dp2[i-1]);
    }
    
    int answer = max(dp1[money_length - 2], dp2[money_length-1]);
    return answer;
}